### PR TITLE
Restore SMP to a working state, enable pervasively on x86_64

### PR DESCRIPTION
**Pull request by andyross**
SMP got kinda rotten across the timer and scheduler rework, as our only SMP platform was a hardware-only device (esp32) with somewhat quirky behavior in hardware test integration and no serious users within the community.

This series fixes a bunch of bugs that crept in, and a few that were never discovered originally (the kernel/smp test itself has been significantly enhanced since).  And it caps it off by making SMP the default on x86_64 (everywhere it works, anyway -- lots of tests aren't smp-safe by design).

Full disclosure: the interprocessor interrupt framework in the "Support synchronous k_thread_abort" patch probably constitutes a new feature from the perspective of the 1.14 code freeze.  But (1) this was much easier than trying to find all the uses of k_thread_abort() throughout the tests and "fix" them to work in SMP (where k_thread_abort() basically didn't work by design) and (2) if you look carefully, I believe it's true that there are ZERO changes in code paths when CONFIG_SMP=n.  It only affects the situation I'm fixing here.

### CODE REVIEW

File: drivers/timer/hpet.c [hunk 1]
```diff
@@ -98,6 +98,14 @@ int z_clock_driver_init(struct device *device)
 	return 0;
 }
 
+void smp_timer_init(void)
```
Comment (dcpleung) [inline drivers/timer/hpet.c]: Might want to wrap this inside "#ifdef CONFIG_SMP"

File: kernel/smp.c [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 #ifdef CONFIG_SMP
 static atomic_t global_lock;
+static atomic_t start_flag;
```
Comment (dcpleung) [inline kernel/smp.c]: I pulled in this patch for my own branch and it failed some sanitychecks. This is defined within CONFIG_SMP, but the usage inside smp_init() below is not inside a #ifdef. So there were errors.

File: drivers/timer/hpet.c (hunk 1 above)
Comment (andyross) [reply drivers/timer/hpet.c]: Not sure I see why; if it's not referenced it won't get pulled into the link.

File: kernel/smp.c (hunk 2 above)
Comment (andyross) [reply kernel/smp.c]: Seems like it is to me and it builds on both SMP=y/n in CI.  Can you post the error?

**Comment by andrewboie** [comment]
Hmm
Do we have bugs open on the tests that explode when SMP is enabled?

**Comment by andyross** [comment]
Threw a DNM tag on this because the PR now how a temporary patch in it to force the use of a particular SDK.   That needs to come out after CI is upgraded and before merge.

And indeed, I should open a bug to audit all those tests.  Most of them aren't really "bugs" -- just tests that didn't expect there to be another CPU.   I mean, the tests/kernel/sched test that I wrote is 100% predicated on the idea of throwing a bunch of threads at a single CPU scheduler and seeing what runs first.

Lots of that kind of thing can be fixed by changing the test code a little to e.g. put a high priority thread on the other CPU for the duration of the specific failing test cases, or by using the SMP affinity API to pin threads, etc...  And I'm sure at least a few of those failures reflect real bug with the SMP or arch layer too.

File: drivers/timer/hpet.c (hunk 1 above)
Comment (dcpleung) [reply drivers/timer/hpet.c]: I was thinking "defined but no used" warning. Come to think of it, it shouldn't matter.

File: kernel/smp.c (hunk 2 above)
Comment (dcpleung) [reply kernel/smp.c]: > zephyr/kernel/smp.c: In function ‘smp_init’:
> zephyr/kernel/smp.c:96: error: ‘start_flag’ undeclared (first use in this function)
> zephyr/kernel/smp.c:96: error: (Each undeclared identifier is reported only once
> zephyr/kernel/smp.c:96: error: for each function it appears in.)
> make[2]: *** [zephyr/kernel/CMakeFiles/kernel.dir/build.make:310: zephyr/kernel/CMakeFiles/kernel.dir/smp.c.obj] Error 2

I was just building `samples/hello_world` with no modification. But this was under XCC.

**Comment by andyross** [comment]
Rebased & fixed the gitlint errors (that were on the DNM patch which can't be merged anyway, but might as well make it show green).  I'm a little at a loss for the build errors where we're overflowing code space on a handful of 32kb platforms.  Those same tests build with the same SDK for me locally.  They're tight, but they fit.  Is something extra turned on for CI builds?

**Comment by andyross** [comment]
OK, figured out the timeouts.  The "modified tests" script is noticing that I've, heh, modified a ton of tests, so it's adding those directories (no shock, they'd get built anyway) **and** a "--all" to the sanitycheck argument, presumably because the tests need to be build-validated on all platforms?  And apparently CI can't currently handle an --all build in the two hours a pull request is given to finish?  Or plausibly the new SDK is slower than the old one by enough to push it over the edge.

@nashif -- what's the best way forward here?

**Comment by andyross** [comment]
Note that the giant "disable SMP in a bunch of tests" patch has been split out into three patches that can be merged before this one (as without this PR, they're noops).  That should fix the CI build time troubles.

**Comment by andrewboie** [comment]
needs rebase.

**Review by dcpleung — APPROVED** [review]

**Review by tyrantlizard77 — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Comment by andyross** [comment]
Oops, pushed the version with the included squash patches.  One sec...

**Pull request merged**